src/main/java/org/apache/commons/lang3/concurrent/AbstractConcurrentInitializer.java

```diff
@@ -151,8 +151,20 @@ public void close() throws E {
      * @return the managed data object
      * @throws E if an error occurs during object creation
      */
+    @SuppressWarnings("unchecked")
     protected T initialize() throws E {
-        return initializer.get();
+        try {
+            return initializer.get();
+        } catch (final Exception e) {
+            // Do this first so we don't pass a RuntimeException into an exception constructor
+            ExceptionUtils.throwUnchecked(e);
+
+            final E typedException = getTypedException(e);
+            if (typedException.getClass().isAssignableFrom(e.getClass())) {
+                throw (E) typedException.getClass().cast(e);
+            }
+            throw typedException;
```

Comment:
This does throw `Exception` if `getTypedException(e)` returns it.

Reply:
It can throw an Exception but it can't throw the `new Exception` object created in some implementations of `getTypedException()`

Reply:
> It can throw an Exception but it can't throw the `new Exception` object created in some implementations of `getTypedException()`

Hm, but that's exactly what happens when in
```
            final E typedException = getTypedException(e);
            if (typedException.getClass().isAssignableFrom(e.getClass())) {
                throw (E) e;
            }
            throw typedException;
```

1. The call to `getTypedException(e);` returns a new `Exception` and is saved in `typedException`.
2. The `if` condition is false
3. Throws the `typedException`, the same new `Exception` returned by `getTypedException(e);`

I am missing something?

Reply:
Unless I've made an error, the `if` condition can never be false for classes where `E` is `Exception`.

If the exception we're processing is an `Error` we will not reach this code block because `ExceptionUtils.throwUnchecked(e);` will throw it. And if the exception is anything else `typedException.getClass().isAssignableFrom(e.getClass())` will return true.

If the above looks correct to you I'll document this in a comment.